src/vm-vcpu-ref/src/x86_64/mptable.rs

```diff
@@ -112,229 +117,266 @@ fn mpf_intel_compute_checksum(v: &mpspec::mpf_intel) -> u8 {
-        + mem::size_of::<MpcBusWrapper>()
-        + mem::size_of::<MpcIntsrcWrapper>() * 16
-        + mem::size_of::<MpcLintsrcWrapper>() * 2
+#[derive(Debug)]
+/// Provides the means to initialize and write to memory a MP (Multi Processor) Table.
+///
+/// The MP Table is created using defaults as provided in the
+/// [MP Spec](https://pdos.csail.mit.edu/6.828/2008/readings/ia32/MPspec.pdf).
+///
+/// At the moment, it only allows configuring the number of vCPUs.
+///
+/// # Example - Write to memory an MP Table for 4 vCPUs.
+///
+/// ```rust
+/// use vm_memory::{GuestAddress, GuestMemoryMmap};
+/// use vm_vcpu_ref::x86_64::mptable::{MpTable, Result};
+///
+/// fn write_mptable() -> Result<()> {
+///     let num_cpus = 4;
+///     let mptable = MpTable::new(num_cpus)?;
+///     let mem: GuestMemoryMmap =
+///         GuestMemoryMmap::from_ranges(&[(GuestAddress(0), 1024 << 20)]).unwrap();
+///     mptable.write(&mem)
+/// }
+///
+/// # write_mptable().unwrap();
+/// ```
+pub struct MpTable {
+    irq_num: u8,
+    cpu_num: u8,
 }
 
-/// Performs setup of the MP table for the given `num_cpus`.
-pub fn setup_mptable<M: GuestMemory>(mem: &M, num_cpus: u8) -> Result<()> {
-    if u32::from(num_cpus) > MAX_SUPPORTED_CPUS {
-        return Err(Error::TooManyCpus);
-    }
-
-    // Used to keep track of the next base pointer into the MP table.
-    let mut base_mp = GuestAddress(MPTABLE_START);
+impl MpTable {
+    /// Creates a new MP Table that can hold `cpu_num`.
+    pub fn new(cpu_num: u8) -> Result<MpTable> {
+        if cpu_num > MAX_SUPPORTED_CPUS {
+            return Err(Error::TooManyCpus);
+        }
 
-    let mp_size = compute_mp_size(num_cpus);
+        Ok(MpTable {
+            irq_num: IRQ_MAX + 1,
+            cpu_num,
+        })
+    }
 
-    let mut checksum: u8 = 0;
-    let ioapicid: u8 = num_cpus + 1;
+    // Returns the size of this MP table based on its configuration.
+    fn size(&self) -> usize {
+        mem::size_of::<MpfIntel>()
+            + mem::size_of::<MpcTable>()
+            + mem::size_of::<MpcCpu>() * (self.cpu_num as usize)
+            + mem::size_of::<MpcIoapic>()
+            + mem::size_of::<MpcBus>()
+            + mem::size_of::<MpcIntsrc>() * (self.irq_num as usize)
+            + mem::size_of::<MpcLintsrc>() * 2
+    }
 
-    // The checked_add here ensures the all of the following base_mp.unchecked_add's will be without
-    // overflow.
-    if let Some(end_mp) = base_mp.checked_add((mp_size - 1) as u64) {
-        if !mem.address_in_range(end_mp) {
-            return Err(Error::NotEnoughMemory);
+    /// Write the MP Table in guest memory.
+    pub fn write<M: GuestMemory>(&self, mem: &M) -> Result<()> {
+        // Used to keep track of the next base pointer into the MP table.
+        let mut base_mp = GuestAddress(MPTABLE_START);
+        let mp_size = self.size();
+        // The checked_add here ensures the all of the following base_mp.unchecked_add's will be without
+        // overflow.
```

Comment:
I would also add a short comment here saying something like: "all the `as u8/u16` below are safe because the numbers from which the conversions are made have small values".